core/src/sumeragi/main_loop.rs

```diff
@@ -889,229 +964,171 @@ pub fn run<F>(
+                        let _enter = span.enter();
+                        match block.revalidate(
+                            &sumeragi.transaction_validator,
+                            &state.wsv,
+                            &state.latest_block_hash,
+                            state.latest_block_height,
+                        ) {
+                            Ok(block) => block,
+                            Err(err) => {
+                                warn!(%err);
+                                continue;
+                            }
                         }
+                    };
 
-                        let voting_block = VotingBlock::new(block.clone());
-                        voting_block_option = Some(voting_block);
+                    for event in Vec::<Event>::from(&block) {
+                        trace!(?event);
+                        sumeragi.events_sender.send(event).unwrap_or(0);
+                    }
 
-                        instant_at_which_we_should_have_committed =
-                            Instant::now() + sumeragi.commit_time;
+                    let valid_signatures = state.current_topology.filter_signatures_by_roles(
+                        &[Role::ValidatingPeer, Role::Leader],
+                        block.verified_signatures(),
+                    );
+                    for sig in &valid_signatures {
+                        block_signature_acc.push((block.hash(), sig.clone()));
                     }
-                    Message::BlockSigned(block_signed) => {
-                        let BlockSigned { hash, signatures } = block_signed;
-
-                        if voting_block_option.is_some()
-                            && hash
-                                != voting_block_option
-                                    .as_ref()
-                                    .expect("Voting block is `Some`")
-                                    .block
-                                    .hash()
-                        {
-                            error!("block signed is not relevant block");
-                            continue;
-                        }
 
-                        let valid_signatures = state.current_topology.filter_signatures_by_roles(
-                            &[Role::ValidatingPeer, Role::Leader],
-                            signatures.verified_by_hash(hash),
-                        );
+                    let voting_block = VotingBlock::new(block.clone());
+                    voting_block_option = Some(voting_block);
 
-                        for sig in valid_signatures {
-                            block_signature_acc.push((hash, sig));
-                        }
+                    instant_at_which_we_should_have_committed =
+                        Instant::now() + sumeragi.commit_time;
+                }
+                Some(Message::BlockSigned(BlockSigned { hash, signatures })) => {
+                    if voting_block_option.is_none()
```

Comment:
Networks are weird. There is a case where we receive the vote before the block. Ideally we should therefore have single "pending vote" slot per foreign peer. The vote gets stored there until we receive the voting block and then we can check it as usual.

It adds a bit of complexity but it is worth it for good runtime behaviour, especially in a public blockchain.

Reply:
Really good point! Have created #2999.